File: DSA/Asmi_Garg/day8_ques1.java
import java.util.Scanner;
import java.util.Stack;

public class EqualStackHeight {

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            System.out.print("Enter the number of cylinders in stack 1:");
            int n1 = scanner.nextInt();
            int[] stack1 = new int[n1];
            System.out.print("Enter the heights of the cylinders in stack 1:");
            for (int i = 0; i < n1; i++) {
                stack1[i] = scanner.nextInt();
            }

            System.out.print("Enter the number of cylinders in stack 2:");
            int n2 = scanner.nextInt();
            int[] stack2 = new int[n2];
            System.out.print("Enter the heights of the cylinders in stack 2:");
            for (int i = 0; i < n2; i++) {
                stack2[i] = scanner.nextInt();
            }

            System.out.print("Enter the number of cylinders in stack 3:");
            int n3 = scanner.nextInt();
            int[] stack3 = new int[n3];
            System.out.print("Enter the heights of the cylinders in stack 3:");
            for (int i = 0; i < n3; i++) {
                stack3[i] = scanner.nextInt();
            }

            System.out.print(
                    "The maximum possible height where all stacks are equal: "
                            + maxEqualHeight(stack1, stack2, stack3));
        }
    }

    public static int maxEqualHeight(int[] stack1, int[] stack2, int[] stack3) {
        Stack<Integer> s1 = convertArrayToStack(stack1);
        Stack<Integer> s2 = convertArrayToStack(stack2);
        Stack<Integer> s3 = convertArrayToStack(stack3);

        int h1 = getHeight(s1);
        int h2 = getHeight(s2);
        int h3 = getHeight(s3);

        while (true) {
            if (s1.isEmpty() || s2.isEmpty() || s3.isEmpty()) {
                return 0;
            }

            if (h1 == h2 && h2 == h3) {
                return h1;
            }

            if (h1 >= h2 && h1 >= h3) {
                h1 -= s1.pop();
            } else if (h2 >= h1 && h2 >= h3) {
                h2 -= s2.pop();
            } else if (h3 >= h1 && h3 >= h2) {
                h3 -= s3.pop();
            }
        }
    }

    private static Stack<Integer> convertArrayToStack(int[] array) {
        Stack<Integer> stack = new Stack<>();
        for (int i = array.length - 1; i >= 0; i--) {
            stack.push(array[i]);
        }
        return stack;
    }

    private static int getHeight(Stack<Integer> stack) {
        int height = 0;
        for (int value : stack) {
            height += value;
        }
        return height;
    }
}

File: DSA/Umang-Singhal/Day8_Ques1.c
#include <stdio.h>
int sum(int stack[], int size) {
    int total = 0;
    for (int i = 0; i < size; i++) {
        total += stack[i];
    }
    return total;
}
int main() {
    int n1, n2, n3;
    printf("Enter the number of cylinders in stack 1: ");
    scanf("%d", &n1);
    int stack1[n1];
    printf("Enter the heights of the cylinders in stack 1: ");
    for (int i = 0; i < n1; i++) {
        scanf("%d", &stack1[i]);
    }
    printf("Enter the number of cylinders in stack 2: ");
    scanf("%d", &n2);
    int stack2[n2];
    printf("Enter the heights of the cylinders in stack 2: ");
    for (int i = 0; i < n2; i++) {
        scanf("%d", &stack2[i]);
    }
    printf("Enter the number of cylinders in stack 3: ");
    scanf("%d", &n3);
    int stack3[n3];
    printf("Enter the heights of the cylinders in stack 3: ");
    for (int i = 0; i < n3; i++) {
        scanf("%d", &stack3[i]);
    }
    int sum1 = sum(stack1, n1);
    int sum2 = sum(stack2, n2);
    int sum3 = sum(stack3, n3);

    int i1 = 0, i2 = 0, i3 = 0;
    while (i1 < n1 && i2 < n2 && i3 < n3) {
        if (sum1 == sum2 && sum2 == sum3) {
            printf("The maximum possible height where all stacks are equal is: %d\n", sum1);
            return 0;
        }
        if (sum1 >= sum2 && sum1 >= sum3) {
            sum1 -= stack1[i1++];
        } else if (sum2 >= sum1 && sum2 >= sum3) {
            sum2 -= stack2[i2++];
        } else if (sum3 >= sum1 && sum3 >= sum2) {
            sum3 -= stack3[i3++];
        }
    }
    printf("The maximum possible height where all stacks are equal is: 0\n");
    return 0;
}
